### PR TITLE
[WIP] Standby cluster implementation

### PR DESCRIPTION
**Pull request by erthalion**
First attempt to implement #657 . Created mostly for discussion and transparency, only the main path is implemented without handling of any corner cases and tests so far. To enable you need to bootstrap a new cluster with:

```yaml
standby_cluster:
  conn_url: postgresql://127.0.0.1:1111/postgres
  replication_slot: patroni
```

### CODE REVIEW

File: patroni/ha.py [hunk 1]
```diff
@@ -187,6 +195,28 @@ def bootstrap(self):
                 return 'trying to ' + msg
             return 'waiting for leader to bootstrap'
 
+    def bootstrap_standby_leader(self):
+        """ If we found 'standby' key in the configuration, we need to bootstrap
+            not a real master, but a 'standby leader', that will take base backup
+            from a remote master and start follow it.
+        """
+        config_value = json.dumps(
```
Comment (CyberDem0n) [inline patroni/ha.py]: There should be nither `/config` nor `/leader` key in DCS until bootstrap is done.

File: patroni/ha.py [hunk 2]
```diff
@@ -725,8 +761,23 @@ def process_unhealthy_cluster(self):
...
                         self.dcs.manual_failover('', '')
                 self.load_cluster_from_dcs()
-                return self.enforce_master_role('acquired session lock as a leader',
-                                                'promoted self to leader by acquiring session lock')
+
+                if self.cluster.is_standby_cluster():
+                    # standby leader disappeared, and this is a healthiest
+                    # replica, so it should become a new standby leader.
+                    # This imply that we need to start following a remote master
+                    msg = 'follow remote master {0}'.format(follow_target.conn_url)
```
Comment (CyberDem0n) [inline patroni/ha.py]: follow_target is not defined here

File: patroni/ha.py [hunk 3]
```diff
@@ -725,8 +761,23 @@ def process_unhealthy_cluster(self):
...
+                    clone_target = self.cluster.get_target_to_follow()
```
Comment (CyberDem0n) [inline patroni/ha.py]: clone_target is unused

File: patroni/ha.py [hunk 4]
```diff
@@ -725,8 +761,23 @@ def process_unhealthy_cluster(self):
...
+                    follow_target = self.cluster.get_target_to_follow()
+
+                    self._async_executor.schedule('follow_remote_master')
+                    self._async_executor.run_async(self.follow, args=(follow_target, msg))
+                    return 'promoted self to a standby leader because i had the session lock'
```
Comment (CyberDem0n) [inline patroni/ha.py]: `promoted self to a standby leader by acquiring session lock`

File: patroni/ha.py [hunk 5]
```diff
@@ -758,8 +809,17 @@ def process_healthy_cluster(self):
                 if msg is not None:
                     return msg
 
-                return self.enforce_master_role('no action.  i am the leader with the lock',
-                                                'promoted self to leader because i had the session lock')
+                if self.cluster.is_standby_cluster():
+                    # in case of standby cluster we don't really need to
```
Comment (CyberDem0n) [inline patroni/ha.py]: What if it was accidentally promoted?
You have to call `self.follow` here.

File: patroni/postgresql.py [hunk 6]
```diff
@@ -1365,7 +1371,9 @@ def follow(self, member, timeout=None):
         if primary_conninfo:
             recovery_params['primary_conninfo'] = primary_conninfo
         if self.use_slots:
-            recovery_params['primary_slot_name'] = slot_name_from_member_name(self.name)
+            required_name = member is not None and member.data.get('replication_slot')
```
Comment (CyberDem0n) [inline patroni/postgresql.py]: In case if we are following the remote cluster and there is no `replication_slot` defined in the configuration `primary_slot_name` shouldn't be defined either.

File: patroni/ha.py [hunk 7]
```diff
@@ -1031,6 +1092,9 @@ def set_start_timeout(self, value):
         self._start_timeout = value
 
     def _run_cycle(self):
+        if self._stop is True:
```
Comment (CyberDem0n) [inline patroni/ha.py]: Doesn't seems to be used.

File: patroni/ha.py [hunk 8]
```diff
@@ -725,8 +761,23 @@ def process_unhealthy_cluster(self):
...
+                if self.cluster.is_standby_cluster():
+                    # standby leader disappeared, and this is a healthiest
+                    # replica, so it should become a new standby leader.
+                    # This imply that we need to start following a remote master
+                    msg = 'follow remote master {0}'.format(follow_target.conn_url)
+                    clone_target = self.cluster.get_target_to_follow()
+                    follow_target = self.cluster.get_target_to_follow()
+
+                    self._async_executor.schedule('follow_remote_master')
+                    self._async_executor.run_async(self.follow, args=(follow_target, msg))
```
Comment (CyberDem0n) [inline patroni/ha.py]: `follow` itself checks recovery.conf and asynchronously calls `state_handler.follow` when it is necessary, therefore no need to use `_async_executor` here

File: patroni/config.py [hunk 9]
```diff
@@ -312,8 +317,16 @@ def _build_effective_configuration(self, dynamic_configuration, local_configurat
         if 'name' not in config and 'name' in pg_config:
             config['name'] = pg_config['name']
 
-        pg_config.update({p: config[p] for p in ('name', 'scope', 'retry_timeout',
-                          'synchronous_mode', 'maximum_lag_on_failover') if p in config})
+        updated_fields = (
+            'name',
```
Comment (CyberDem0n) [inline patroni/config.py]: We actually need only first 3 here...

File: patroni/dcs/__init__.py [hunk 10]
```diff
@@ -426,6 +440,10 @@ def leader_optime_path(self):
     def sync_path(self):
         return self.client_path(self._SYNC)
 
+    @property
+    def standby_path(self):
```
Comment (CyberDem0n) [inline patroni/dcs/__init__.py]: There is no standby key in DCS.

File: patroni/config.py [hunk 11]
```diff
@@ -88,6 +89,10 @@ def config_file(self):
     def dynamic_configuration(self):
         return deepcopy(self._dynamic_configuration)
 
+    @property
+    def is_standby_cluster(self):
```
Comment (CyberDem0n) [inline patroni/config.py]: There is no need to have `is_standby_cluster` property here.

The only reason why we keep information about `standby_cluster` in the `Config()` is just to be able to dump into `patroni.dynamic.json`. It makes it possible to restore it if somebody accidentally wipes DCS out.

File: patroni/ha.py (hunk 4 above)
Comment (erthalion) [reply patroni/ha.py]: This branch is within `if self.acquire_lock()`, isn't it enough?

File: patroni/config.py (hunk 11 above)
Comment (erthalion) [reply patroni/config.py]: There is such need - because we need to get this info directly from config file before bootstrap. It's not really good to read it directly from `_dynamic_configuration`, so I've created a property.

File: patroni/ha.py [hunk 12]
```diff
@@ -252,12 +278,18 @@ def recover(self):
     def _get_node_to_follow(self, cluster):
         # determine the node to follow. If replicatefrom tag is set,
         # try to follow the node mentioned there, otherwise, follow the leader.
-        if not self.patroni.replicatefrom or self.patroni.replicatefrom == self.state_handler.name:
-            node_to_follow = cluster.leader
-        else:
+        is_leader = self.cluster.leader and self.state_handler.name == self.cluster.leader.name
+
+        if self.patroni.replicatefrom and self.patroni.replicatefrom != self.state_handler.name:
```
Comment (CyberDem0n) [inline patroni/ha.py]: What if node is the leader of standby cluster and replicatefrom tag is defined?
We will get a replication loop :)

File: patroni/ha.py [hunk 13]
```diff
@@ -397,6 +429,15 @@ def update_cluster_history(self):
                         line.append(cluster_history[line[0]][3])
                 self.dcs.set_history_value(json.dumps(history, separators=(',', ':')))
 
+    def enforce_follow_remote_master(self):
+        self.state_handler.set_role('standby_leader')
+        follow_target = self.cluster.get_target_to_follow()
+        message = 'follow remote master {0}'.format(follow_target.conn_url)
+
+        self._async_executor.schedule('follow_remote_master')
+        self._async_executor.run_async(self.follow, args=(follow_target, message))
```
Comment (CyberDem0n) [inline patroni/ha.py]: It seems that you wanted to call `Postgresql.follow`, but it also has a different signature.

I absolutely agree that we shouldn't touch existing `Ha.follow`, but at the same time `enforce_follow_remote_master` should implement most of the functionality of the `Ha.follow`.

Also `Ha._handle_rewind` should do a rewind from remote master, not from the cluster leader.

File: patroni/ha.py [hunk 14]
```diff
@@ -758,8 +809,17 @@ def process_healthy_cluster(self):
...
                     return msg
 
-                return self.enforce_master_role('no action.  i am the leader with the lock',
-                                                'promoted self to leader because i had the session lock')
+                if self.cluster.is_standby_cluster():
+                    # in case of standby cluster we don't really need to
+                    # enforce anything, since the leader is not a master.
+                    # So just remind the role.
+                    self.enforce_follow_remote_master()
+                    return 'no action. I am the standby leader with the lock'
```
Comment (CyberDem0n) [inline patroni/ha.py]: It might happen that postgres will be restarted due to change in `recovery.conf`, but it will still report "no action".

File: patroni/postgresql.py [hunk 15]
```diff
@@ -1365,7 +1371,10 @@ def follow(self, member, timeout=None):
...
+            name = required_name or slot_name_from_member_name(self.name)
```
Comment (CyberDem0n) [inline patroni/postgresql.py]: Unfortunately it doesn't fix a problem.
Lets assume that we want to follow the remote cluster without using replication slots.
So we have no `replication_slot defined` in the `standby_cluster`, but according to this line it will set `primary_slot_name` to the `slot_name_from_member_name(self.name)`.

Actually the same function should be applied to the `required_name`

File: patroni/config.py [hunk 16]
```diff
@@ -45,6 +45,7 @@ class Config(object):
         'master_start_timeout': 300,
         'synchronous_mode': False,
         'synchronous_mode_strict': False,
+        'standby_cluster': '',
```
Comment (CyberDem0n) [inline patroni/config.py]: I've though a little bit about format of `standby_cluster` configuration and would suggest to follow postgres naming conventions here: https://www.postgresql.org/docs/10/static/recovery-config.html
```python
{
    'host': 'localhost',
    'port': 12345,
    'primary_slot_name': 'foo-bar',
    'restore_command': 'true',
    'archive_cleanup_command': ''
    'recovery_min_apply_delay': '10min'
}
```

We can assume that this is a standby cluster when either `host` or `restore_command` are defined.

File: patroni/postgresql.py (hunk 15 above)
Comment (erthalion) [reply patroni/postgresql.py]: Hm...looks like it means that this function should know whether this particular member is a part of standby cluster or not, am I right?

File: patroni/config.py [hunk 17]
```diff
@@ -45,6 +46,14 @@ class Config(object):
         'master_start_timeout': 300,
         'synchronous_mode': False,
         'synchronous_mode_strict': False,
+        'standby_cluster': {
+            'host': 'localhost',
```
Comment (erthalion) [inline patroni/config.py]: Empty strings for values

File: patroni/ha.py [hunk 18]
```diff
@@ -1201,3 +1265,21 @@ def wakeup(self):
...
+        config = config or (self.config is not None and self.config.data)
+
+        if config and config.get('standby_cluster'):
+            cluster_params = config.get('standby_cluster')
+            return Member(None, 'remote_master', None, {
+                'conn_kwargs': {
+                    "host": cluster_params.get('host'),
+                    "port": cluster_params.get('port'),
+                },
+                'primary_slot_name': cluster_params['primary_slot_name'],
```
Comment (CyberDem0n) [inline patroni/ha.py]: What if `primary_slot_name` is not specified?

File: patroni/ha.py [hunk 19]
```diff
@@ -1201,3 +1265,21 @@ def wakeup(self):
...
+                'no_replication_slot': False,
```
Comment (CyberDem0n) [inline patroni/ha.py]: Must be `True` if there is no `primary_slot_name` defined.

File: patroni/ha.py [hunk 20]
```diff
@@ -1201,3 +1265,21 @@ def wakeup(self):
...
+    def get_target_to_follow(self, config):
+        """ In case of standby cluster this will tel us from which remote
+            master to stream. Config can be both patroni config or
+            cluster.config.data
+        """
+        config = config or (self.config is not None and self.config.data)
+
+        if config and config.get('standby_cluster'):
+            cluster_params = config.get('standby_cluster')
+            return Member(None, 'remote_master', None, {
```
Comment (CyberDem0n) [inline patroni/ha.py]: What is one of the nodes in the cluster has the name `remote_master`?

File: features/steps/standby_cluster.py [hunk 21]
```diff
@@ -0,0 +1,25 @@
...
+def start_patroni_stanby_cluster(context, name, cluster_name, name2):
+    port = context.pctl._processes[name2]._connkwargs.get('port')
+    return context.pctl.start(name, custom_config={
+        "scope": cluster_name,
+        "bootstrap": {
+            "dcs": {
+                "standby_cluster" :{
+                    "host": "localhost",
+                    "port": port,
+                    "primary_slot_name": "postgresql1",
```
Comment (CyberDem0n) [inline features/steps/standby_cluster.py]: Since there is no replication slot on the `postgresql0` it won't be able to connect and stream...

File: features/standby_cluster.feature [hunk 22]
```diff
@@ -0,0 +1,16 @@
...
+    Then postgres1 is a leader of batman1 after 10 seconds
+    And postgres1 role is the standby_leader after 24 seconds
+    And table foo is present on postgres1 after 20 seconds
+    When I start postgres2 in a cluster batman1
+    Then postgres2 role is the replica after 24 seconds
+    And table foo is present on postgres2 after 20 seconds
+
+  Scenario: check failover
+    When I kill postmaster on postgres1
+    Then postgres2 role is the standby_leader after 24 seconds
```
Comment (CyberDem0n) [inline features/standby_cluster.feature]: At the end of the day it will call: https://github.com/zalando/patroni/blob/master/features/environment.py#L219, which executes `pg_is_in_recovery()` and treats roles as `'primary'` and `not 'primary'`

File: patroni/common_config.py [hunk 23]
```diff
@@ -0,0 +1,9 @@
+def is_standby_cluster(config):
+    """ Check whether or not provided configuration describes a standby cluster.
+        Config can be both patroni config or cluster.config.data
+    """
+    return config is not None and (
```
Comment (CyberDem0n) [inline patroni/common_config.py]:
```python
isinstance(config, dict) and (config.get('host') or config.get('port') or config.get('restore_command'))
```

Because:
```python
In [1]: '' is not None
Out[1]: True
```

File: patroni/ha.py [hunk 24]
```diff
@@ -1201,3 +1266,25 @@ def wakeup(self):
...
+        if config and config.get('standby_cluster'):
+            cluster_params = config.get('standby_cluster')
+            name = self.unique_name_for_remote_master()
+            return Member(None, name, None, {
+                'conn_kwargs': {
+                    "host": cluster_params.get('host'),
+                    "port": cluster_params.get('port'),
+                },
+                'primary_slot_name': cluster_params.get('primary_slot_name'),
+                'no_replication_slot': not hasattr(cluster_params, 'primary_slot_name'),
```
Comment (CyberDem0n) [inline patroni/ha.py]: `hasattr` will always return False in this case

File: patroni/ha.py [hunk 25]
```diff
@@ -733,8 +775,18 @@ def process_unhealthy_cluster(self):
...
                         self.dcs.manual_failover('', '')
                 self.load_cluster_from_dcs()
-                return self.enforce_master_role('acquired session lock as a leader',
-                                                'promoted self to leader by acquiring session lock')
+
+                if self.cluster.is_standby_cluster():
+                    # standby leader disappeared, and this is a healthiest
+                    # replica, so it should become a new standby leader.
+                    # This imply that we need to start following a remote master
+                    self.enforce_follow_remote_master()
```
Comment (alexeyklyukin) [inline patroni/ha.py]: I'd pass the message at line 784 to the enforce_follow_remote_master and return its result, which should be either  a message that the actual Postgres master demoted, or that we have promoted ourself to the standby leader and following the remote leader. Right now, the message from line 448 is simply discarded here.

File: patroni/ha.py [hunk 26]
```diff
@@ -766,8 +818,17 @@ def process_healthy_cluster(self):
                 if msg is not None:
                     return msg
 
-                return self.enforce_master_role('no action.  i am the leader with the lock',
-                                                'promoted self to leader because i had the session lock')
+                if self.cluster.is_standby_cluster():
+                    # in case of standby cluster we don't really need to
+                    # enforce anything, since the leader is not a master.
+                    # So just remind the role.
+                    self.enforce_follow_remote_master()
```
Comment (alexeyklyukin) [inline patroni/ha.py]: same consideration: make enforce_follow_remote_master return you the lock you need.

File: patroni/ha.py [hunk 27]
```diff
@@ -1201,3 +1266,25 @@ def wakeup(self):
         no "active" leader watch request in progress.
         This usually happens on the master or if the node is running async action"""
         self.dcs.event.set()
+
+    def unique_name_for_remote_master(self):
+        return 'remote_master:{}'.format(uuid.uuid1())
+
+    def get_target_to_follow(self, config):
```
Comment (alexeyklyukin) [inline patroni/ha.py]: Too generic name. It seems to be used only in the standby_cluster scenario and only for figuring out the node that is the remote master. So why not name it get_remote_master?

Comment (alexeyklyukin) [reply patroni/ha.py]: Besides, it is used not only for finding node to follow, but also for finding the one to clone from.

File: patroni/ha.py [hunk 28]
```diff
@@ -52,6 +54,7 @@ def failover_limitation(self):
 class Ha(object):
 
     def __init__(self, patroni):
+        self._stop = False
```
Comment (CyberDem0n) [inline patroni/ha.py]: I would still insist on removing it

File: features/standby_cluster.feature [hunk 29]
```diff
@@ -3,10 +3,9 @@ Feature: standby cluster
   Scenario: check replication of a single table in a standby cluster
     Given I start postgres0 without slots sync
     And I create a replication slot postgres1 on postgres0
-    And I create a role replicator on postgres0
     And I start postgres1 in a standby cluster batman1 as a clone of postgres0
     Then postgres1 is a leader of batman1 after 10 seconds
-    And postgres1 role is the standby_leader after 24 seconds
+    And DCS for batman1 has leader=postgres1 after 20 seconds
```
Comment (CyberDem0n) [inline features/standby_cluster.feature]: https://github.com/zalando/patroni/blob/master/features/custom_bootstrap.feature#L9 -- is already doing exactly what you need

File: docs/SETTINGS.rst [hunk 30]
```diff
@@ -25,6 +25,14 @@ Bootstrap configuration
         -  **use\_slots**: whether or not to use replication_slots. Must be False for PostgreSQL 9.3. You should comment out max_replication_slots before it becomes ineligible for leader status.
         -  **recovery\_conf**: additional configuration settings written to recovery.conf when configuring follower. 
         -  **parameters**: list of configuration settings for Postgres. Many of these are required for replication to work.
+    -  **standby\_cluster**: if this section is defined, we want to bootstrap a standby cluster.
+        -  **host**: an address of remote master
+        -  **port**: a port of remote master
+        -  **primary\_slot\_name**: which slot on the remote master to use for replication
```
Comment (CyberDem0n) [inline docs/SETTINGS.rst]: Would be good to mention that this is an optional parameter.

File: docs/SETTINGS.rst [hunk 31]
```diff
@@ -25,6 +25,14 @@ Bootstrap configuration
...
+        -  **create\_replica\_methods**: an ordered list of methods that can be used to bootstrap standby leader from the remote master, can be different from the list defined in :ref:`postgresql_settings`
+        -  **restore\_command**: command to restore WAL records from the remote master to standby leader, can be different from the list defined in :ref:`postgresql_settings`
+        -  **archive\_cleanup\_command**: cleanup command for standby leader
```
Comment (CyberDem0n) [inline docs/SETTINGS.rst]: `archive_cleanup_command` and `recovery_min_apply_delay` are not yet supported:
* https://github.com/zalando/patroni/pull/679/files#diff-57c53e988b775ee2902488fb33c1f433R1295
* https://github.com/zalando/patroni/pull/679/files#diff-d8c1582337c8426ef750420dc6b76949R1437

File: docs/replica_bootstrap.rst [hunk 32]
```diff
@@ -129,3 +129,36 @@ and
...
+
+Another available option is to run a "standby cluster", that contains only of
+standby nodes replicating from some remote master. This type of clusters has:
+
+* "standby leader", that behaves pretty much like a regular cluster leader,
+  except it replicates from a remote master.
+
+* cascade replicas, that are replicating from standby leader.
+
+Standby leader holds and updates a leader lock in DCS, and if it will
```
Comment (bradnicholson) [inline docs/replica_bootstrap.rst]: Would read cleaner as:

```
The standby leader holds and updates a leader lock in DCS. If the leader lock expires, cascade replicas will perform an election to choose another leader from the standbys.
```

Comment (erthalion) [reply docs/replica_bootstrap.rst]: Yes, sounds good, thanks.

File: docs/replica_bootstrap.rst [hunk 33]
```diff
@@ -129,3 +129,36 @@ and
...
+"standby mode", and after it was detached to function as a normal cluster.
+
+To configure such cluster you need to specify the section ``standby_cluster``
+in a patroni configuration:
+
+.. code:: YAML
+
+    bootstrap:
+        dcs:
+            postgresql:
```
Comment (CyberDem0n) [inline docs/replica_bootstrap.rst]: `standby_cluster` is not inside `postgresql:`

File: patroni/config.py [hunk 34]
```diff
@@ -183,6 +197,13 @@ def _safe_copy_dynamic_configuration(self, dynamic_configuration):
                         config['postgresql'][name].update(self._process_postgresql_parameters(value))
                     elif name not in ('connect_address', 'listen', 'data_dir', 'pgpass', 'authentication'):
                         config['postgresql'][name] = deepcopy(value)
+            elif name == 'standby_cluster':
+                allowed_keys = self.__DEFAULT_CONFIG['standby_cluster'].keys()
+                expected = {
+                    k: v for k, v in (value or {}).items()
+                    if (k in allowed_keys and isinstance(v, six.string_types))
```
Comment (redbaron) [inline patroni/config.py]: Before trying it, I was looking at how this standby_cluster feature is implemented and just spotted that this one drops `create_replica_methods` which is a list, is it intentional?
